### ISSUE TITLE
Wrong result while training using GPU

### ISSUE THREAD

**Issue by ln1equals0** [torch/nn#1074]
Hi, we have a simple CNN network, which works fine on CPU, but since we have a new GPU installed (TITAN X), we want to train the network using GPU, but it gives us wrong results.

This is the code for training on CPU, it works fine
<img width="1174" alt="2016-12-21 2 17 50" src="https://cloud.githubusercontent.com/assets/10298708/21390665/62dc1262-c788-11e6-9788-8d399175bb10.png">
<img width="1186" alt="2016-12-21 2 17 07" src="https://cloud.githubusercontent.com/assets/10298708/21390667/6596c1f0-c788-11e6-96cf-6a152c9ba943.png">
<img width="1112" alt="2016-12-21 2 12 26" src="https://cloud.githubusercontent.com/assets/10298708/21390605/07b8c92a-c788-11e6-96ec-02e5fc9e4525.png">
<img width="1091" alt="2016-12-21 2 12 45" src="https://cloud.githubusercontent.com/assets/10298708/21390609/0a6d6f86-c788-11e6-9560-db610b162960.png">

This is the result, I made it display the average loss every 100 iterations.
![screenshot from 2016-12-21 14_06_18](https://cloud.githubusercontent.com/assets/10298708/21390700/86e1bc98-c788-11e6-860d-09f203886f9c.png)

<br>
<br>
So then I modified it and try to make it work on GPU, this are the changes I made
<img width="1203" alt="2016-12-21 2 22 01" src="https://cloud.githubusercontent.com/assets/10298708/21390797/f03bb18a-c788-11e6-8ba4-3b430cf01cdb.png">
<img width="1279" alt="2016-12-21 2 22 15" src="https://cloud.githubusercontent.com/assets/10298708/21390806/f9a0df3e-c788-11e6-8a96-6713c8a3a644.png">

It is indeed much faster, but the result is totally wrong, you can see the loss of first 100 - 300 iterations  became 3.4+, and it doesn't converge at all, the loss will stay at 3.3-3.4 till the end of training. 
![screenshot from 2016-12-21 14_05_58](https://cloud.githubusercontent.com/assets/10298708/21390836/1cb81a0a-c789-11e6-9b6c-e583b206e211.png)

Did I miss something? How can I make the training work on GPU. Thanks

**Comment by soumith** [comment]
the default tensor on CPU is DoubleTensor, and on GPU it is FloatTensor. That's one thing you might watch out for when using Sigmoid + BCECriterion, because the criterion can get unstable in the backward at Float precision. Alternatively, you can change your output to just be a 2-element vector, and use an nn.CrossEntropyCriterion
